src/solver.cpp

```diff
@@ -213,154 +212,173 @@ void Solver::getDiagnosisB(const int it, ReturnData *rdata, int which) const {
-             if (status != AMICI_SUCCESS) return;
-             
-             status = SetBandJacFnB(ami_mem, bwd->getwhich());
-             if (status != AMICI_SUCCESS) return;
-             #else*/
-            throw AmiException("Solver currently not supported!");
-            /* #endif*/
-            
-        case LinearSolver::diag:
-            diagB(which);
-            setDenseJacFnB(which);
-            break;
-            
-            /* ITERATIVE SOLVERS */
-            
-        case LinearSolver::SPGMR:
-            spgmrB(which, PREC_NONE, SUNSPGMR_MAXL_DEFAULT);
-            setJacTimesVecFnB(which);
-            break;
-            
-        case LinearSolver::SPBCG:
-            spbcgB(which, PREC_NONE, SUNSPBCGS_MAXL_DEFAULT);
-            setJacTimesVecFnB(which);
-            break;
-            
-        case LinearSolver::SPTFQMR:
-            sptfqmrB(which, PREC_NONE, SUNSPTFQMR_MAXL_DEFAULT);
-            setJacTimesVecFnB(which);
+    /* DIRECT SOLVERS */
+    case LinearSolver::dense:
+        linearSolverB = std::make_unique<SUNLinSolDense>(*x);
+        setLinearSolverB(which);
+        setDenseJacFnB(which);
+        break;
+
+    case LinearSolver::band:
+        linearSolverB = std::make_unique<SUNLinSolBand>(*x, model->ubw, model->lbw);
+        setLinearSolverB(which);
+        setBandJacFnB(which);
+        break;
+
+    case LinearSolver::LAPACKDense:
+        throw AmiException("Solver currently not supported!");
+
+    case LinearSolver::LAPACKBand:
+        throw AmiException("Solver currently not supported!");
+
+    case LinearSolver::diag:
+        diagB(which);
+        setDenseJacFnB(which);
+        break;
+
+        /* ITERATIVE SOLVERS */
+
+    case LinearSolver::SPGMR:
+        linearSolverB = std::make_unique<SUNLinSolSPGMR>(*x, PREC_NONE, SUNSPGMR_MAXL_DEFAULT);
+        setLinearSolverB(which);
+        setJacTimesVecFnB(which);
+        break;
+
+    case LinearSolver::SPBCG:
+        linearSolverB = std::make_unique<SUNLinSolSPBCGS>(*x, PREC_NONE, SUNSPBCGS_MAXL_DEFAULT);
+        setLinearSolverB(which);
+        setJacTimesVecFnB(which);
+        break;
+
+    case LinearSolver::SPTFQMR:
+        linearSolverB = std::make_unique<SUNLinSolSPTFQMR>(*x, PREC_NONE, SUNSPTFQMR_MAXL_DEFAULT);
+        setLinearSolverB(which);
+        setJacTimesVecFnB(which);
+        break;
+
+        /* SPARSE SOLVERS */
+
+    case LinearSolver::KLU:
+        linearSolverB = std::make_unique<SUNLinSolKLU>(*x, model->nnz, CSC_MAT, getStateOrdering());
+        setLinearSolverB(which);
+        setSparseJacFnB(which);
+        break;
+
+    default:
+        throw AmiException("Invalid choice of solver: %d", static_cast<int>(linsol));
+    }
+}
+
+void Solver::initializeNonLinearSolverB(AmiVector *xB, const int which, Model *model)
+{
+    switch(iter) {
+    case NonlinearSolverIteration::newton:
+        switch(ism) {
+        case InternalSensitivityMethod::staggered:
```

Comment:
We never use backwards integration + sensitivities, so this should not be necessary.